Add support for desktop notifications

gpanders opened the pull request:
Fixes: https://github.com/mitchellh/ghostty/issues/612

This PR adds support to core for OSC 9 and OSC 777 for sending desktop notifications from terminal applications. For now, the actual notifications are only implemented for macOS, but all of the core plumbing is there to make it work with GTK (it just needs updates to the apprt surface).

gpanders committed:
core: support OSC 9 and OSC 777 for showing desktop notifications

gpanders committed:
macos: implement desktop notifications

mitchellh reviewed (commented):
This looks really good! Thanks for doing this. I think there are a few issues to chase down.

The big one I found is the nature of packing the pointer address into the user info of a notification. If you trigger a notification, then close the surface, then click the notification, you get a segfault (maybe, but in any case you're accessing illegal memory). That's an issue.

The second, I'm not sure if this is just a problem with my config. After approving notifications, notifications aren't showing up until I click the control center area. If I change my system settings to "alert", they show up, but it defaulted to "banners" for some reason. I can see you request alert privileges. 

The third is more minor, but it's unclear that a notification is from a terminal program and looks officially like Ghostty. That freaks me out a bit, I'm unsure if there is a way we can word the notification title or body in some way to better signal this...

gpanders committed:
macos: remove optional from Ghostty.AppState.surfaceUserdata

The return value of takeUnretainedValue() is non-optional, so this
function was never _actually_ returning an optional value, so the guard
clauses sprinkled throughout were unnecessary.

gpanders committed:
core: use arrays instead of WriteReq for desktop notifications

rockorager committed:
gtk: implement desktop notifications
